gatsby/content/blog/2021/12/2021-12-10-sydent-typing-2.mdx

```diff
@@ -0,0 +1,384 @@
+- [`Optional`](https://docs.python.org/3/library/typing.html?highlight=typing%20generic#typing.Optional): I use this all the time. It's always good to make your `None`s explicit!
+- [`NewType`](https://docs.python.org/3/library/typing.html?highlight=typing%20generic#typing.NewType): I haven't had a chance to use it much. As I understand it, it's a way to define a "strong typedef". For example, we can use it to distinguish lengths from durations, even if they're both represented by a `float` at runtime.
+
+I want to call out two parts of `typing` in particular:
+
+
+#### [overload](https://docs.python.org/3/library/typing.html?highlight=typing%20generic#typing.overload)
+
+`overload` is a way to provide extra information about a function depending on how it's called. For instance, consider this function which takes a `str` or `bytes` as input and returns its uppercase version.
+
+```python
+def upper(x: Union[bytes, str]) -> Union[bytes, str]:
+    return x.upper()
+```
+
+The problem with this annotation is that we'll have to check at every call site to see if the return value was a `bytes` or a `str` object. But we know that uppercasing a `str` gives us a `str`, and uppercasing a `bytes` gives us a `bytes.` We can use `overload` to express this.
+
+```python
+from typing import overload
+
+@overload
+def upper(x: str) -> str: ...
+
+@overload
+def upper(x: bytes) -> bytes: ...
+
+def upper(x: Union[bytes, str]) -> Union[bytes, str]:
+    return x.upper()
+```
+
+The first two `@overload` definitions are like stubs: they're purely used for their annotations. The actual runtime implementation is specified at the end, undecorated. (**NB**: this specific example is better expressed without overloads, by using [`AnyStr`](https://docs.python.org/3/library/typing.html?highlight=typing%20generic#typing.AnyStr).)
+
+I was really impressed to see how `mypy` could use this overloading information. Here's an example:
+
+```python
+from typing import overload, Literal
+
+@overload
+def f(x: int) -> Literal[1]: ...
+
+@overload
+def f(x: str) -> Literal[2]: ...
+
+def f(x: object) -> int:
+    if isinstance(x, int):
+        return 1
+    elif isinstance(x, str):
+        return 2
+    return 3
+
+if f(10) == 2:
+    print("potato")
+```
+
+We can see that `f(10) == 1` and so the equality is always `False`: we'll never print the word "potato". Mypy can reason through this too, if we ask it nicely.
+
+```
+$ mypy example.py
+Success: no issues found in 1 source file
+
+$ mypy --strict-equality --warn-unreachable example.py
+example.py:16: error: Non-overlapping equality check (left operand type: "Literal[1]", right operand type: "Literal[2]")
+example.py:17: error: Statement is unreachable
+Found 2 errors in 1 file (checked 1 source file)
+```
+
+#### [TypedDict](https://docs.python.org/3/library/typing.html?highlight=typeddict#typing.TypedDict)
+
+I mentioned this earlier when talking about the missing `await` bug.
```

Comment:
s/earlier/in last week's blog post/

(with link?)

Reply:
Thanks, this is 6b4910f1